pkg/boot/uefi/uefi.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uefi
+
+import (
+	"bytes"
+	"debug/pe"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+
+	"github.com/u-root/u-root/pkg/acpi"
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/boot/kexec"
+	"github.com/u-root/u-root/pkg/smbios"
+)
+
+const fvEntryImageOffset int64 = 0xA0
+
+var bootExecute = boot.Execute
+var kexecLoad = kexec.Load
+var kexecParseMemoryMap = kexec.ParseMemoryMap
+var getRSDP = acpi.GetRSDP
+
+func getSMBIOSBaseInternal() (int64, int64, error) {
+	base, size, err := smbios.GetSMBIOSBaseEFI()
+	if err != nil {
+		base, size, err = smbios.GetSMBIOSBaseLegacy()
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+	return base, size, nil
+}
+
+var getSMBIOSBase = getSMBIOSBaseInternal
+
+// Serial port setting in Linuxboot.h
+const (
+	SerialPortTypeIO   = 1
+	SerialPortTypeMMIO = 2
+)
+
+// SerialPortConfig defines debug port configuration
+type SerialPortConfig struct {
+	Type        uint32
+	BaseAddr    uint32
+	Baud        uint32
+	RegWidth    uint32
+	InputHertz  uint32
+	UartPciAddr uint32
+}
+
+// Current Config Version: 1
+const PayloadConfigVersion = 1
+
+type payloadConfig struct {
+	Version             uint64
+	AcpiBase            uint64
+	AcpiSize            uint64
+	SmbiosBase          uint64
+	SmbiosSize          uint64
+	SerialConfig        SerialPortConfig
+	NumMemoryMapEntries uint32
+}
+
+// FvImage is a structure for loading a firmware volume
+type FvImage struct {
+	name         string
+	mem          kexec.Memory
+	entryAddress uintptr
+	ImageBase    uintptr
+	SerialConfig SerialPortConfig
+}
+
+func checkFvAndGetEntryPoint(name string) (uintptr, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	// Entry PE image for UEFI Payload is 0xA0 offset from FV head.
+	pef := io.NewSectionReader(f, fvEntryImageOffset, math.MaxInt64)
+	pf, err := pe.NewFile(pef)
+	if err != nil {
+		return 0, err
+	}
+	op64, ok := pf.OptionalHeader.(*pe.OptionalHeader64)
+	if !ok {
+		return 0, fmt.Errorf("It is not OptionalHeader64")
+	}
+	return uintptr(fvEntryImageOffset) + uintptr(op64.AddressOfEntryPoint), nil
+}
+
+// New loads the file and return FvImage stucture if entry image is found
+func New(n string) (*FvImage, error) {
+	entry, err := checkFvAndGetEntryPoint(n)
+	if err != nil {
+		return nil, err
+	}
+	return &FvImage{name: n, mem: kexec.Memory{}, entryAddress: entry}, nil
+}
+
+// Reserved 64kb for passing params
+const uefiPayloadConfigSize = 0x10000
+
+// Load loads fimware volume payload and boot the the payload
+func (fv *FvImage) Load(verbose bool) error {
+	// Install payload
+	dat, err := ioutil.ReadFile(fv.name)
+	if err != nil {
+		return err
+	}
+	fv.mem.Segments.Insert(kexec.NewSegment(dat, kexec.Range{Start: fv.ImageBase, Size: uint(len(dat))}))
```

Comment:
FYI: You've probably noticed at this point that we insert segments without really checking that the memory is unused. We also merge them incorrectly (https://github.com/u-root/u-root/pull/1847 to address that...).

It'd be great if you could also work on addressing these issues beyond the PoC (not necessary now, of course).

Reply:
I will take a look.